django/db/models/fields/related.py

```diff
@@ -615,60 +615,56 @@ def _check_unique_target(self):
         if not self.foreign_related_fields:
             return []
 
-        unique_foreign_fields = {
-            frozenset([f.name])
-            for f in self.remote_field.model._meta.get_fields()
-            if getattr(f, "unique", False)
-        }
-        unique_foreign_fields.update(
-            {frozenset(ut) for ut in self.remote_field.model._meta.unique_together}
-        )
-        unique_foreign_fields.update(
-            {
-                frozenset(uc.fields)
-                for uc in self.remote_field.model._meta.total_unique_constraints
-            }
+        has_unique_constraint = any(
+            rel_field.unique for rel_field in self.foreign_related_fields
         )
-        foreign_fields = {f.name for f in self.foreign_related_fields}
-        has_unique_constraint = any(u <= foreign_fields for u in unique_foreign_fields)
-
-        if not has_unique_constraint and len(self.foreign_related_fields) > 1:
-            field_combination = ", ".join(
-                "'%s'" % rel_field.name for rel_field in self.foreign_related_fields
+        if not has_unique_constraint:
+            foreign_fields = {f.name for f in self.foreign_related_fields}
+            remote_opts = self.remote_field.model._meta
+            has_unique_constraint = any(
+                frozenset(ut) <= foreign_fields for ut in remote_opts.unique_together
+            ) or any(
+                frozenset(uc.fields) <= foreign_fields
+                for uc in remote_opts.total_unique_constraints
             )
-            model_name = self.remote_field.model.__name__
-            return [
-                checks.Error(
-                    "No subset of the fields %s on model '%s' is unique."
-                    % (field_combination, model_name),
-                    hint=(
-                        "Mark a single field as unique=True or add a set of "
-                        "fields to a unique constraint (via unique_together "
-                        "or a UniqueConstraint (without condition) in the "
-                        "model Meta.constraints)."
-                    ),
-                    obj=self,
-                    id="fields.E310",
-                )
-            ]
-        elif not has_unique_constraint:
-            field_name = self.foreign_related_fields[0].name
-            model_name = self.remote_field.model.__name__
-            return [
-                checks.Error(
-                    "'%s.%s' must be unique because it is referenced by "
-                    "a foreign key." % (model_name, field_name),
-                    hint=(
-                        "Add unique=True to this field or add a "
-                        "UniqueConstraint (without condition) in the model "
-                        "Meta.constraints."
-                    ),
-                    obj=self,
-                    id="fields.E311",
+
+        if not has_unique_constraint:
+            if len(self.foreign_related_fields) > 1:
```

Comment:
Here I only rearranged the conditional to check `has_unqiue_constraint` once, mostly for clarity. Use “hide whitespace” when reviewing to clarify.